osidb/migrations/0081_migrate_flawacknowledgment_flawmeta.py

```diff
@@ -0,0 +1,55 @@
+# Written manually on 2023-07-07
+
+from django.db import migrations
+from ast import literal_eval
+
+
+def forwards_func(apps, schema_editor):
+    FlawMeta = apps.get_model("osidb", "FlawMeta")
+    FlawAcknowledgment = apps.get_model("osidb", "FlawAcknowledgment")
+    for meta_ack in FlawMeta.objects.filter(type="ACKNOWLEDGMENT"):
+        flaw = meta_ack.flaw
+        name = meta_ack.meta_attr["name"]
+        # Convert None to empty string.
+        affiliation = meta_ack.meta_attr.get("affiliation") or ""
+        # hstore holds bolean values as strings containing True|False
+        # so we need to explicitly convert it to the bolean value
+        from_upstream = literal_eval(meta_ack.meta_attr["from_upstream"])
+
+        # FlawAcknowledgment that already exist are skipped because they
+        # are exactly as up-to-date as FlawMeta ACKNOWLEDGMENT.
+
+        if not FlawAcknowledgment.objects.filter(
+            flaw=flaw, name=name, affiliation=affiliation
+        ).exists():
+            FlawAcknowledgment.objects.create(
```

Comment:
Could we use [bulk_create](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) instead? In production this will be a lot of data and the migration could take too long, potentially making the deployment fail

Reply:
I did consider it, and I think it would require one to be careful about consumed memory. That quickly got me into a rabbit hole of using a generator instead of a list and setting `batch_size` for bulk_create. Could be done, but potentially complicated for me. I agree with your suggestion, but I'm not skilled enough to pull it off quickly and be otherwise productive.

So, I tested the current approach on production data and it took only a few seconds.

Theoretically, there can be an arbitrary number of FlawAcknowledgment instances, but I think in reality the number has a low upper bound.

IMO, the migration will be executed only in these two situations:

- On existing osidb instances in the year 2023 -> just a few seconds.
- On new osidb instances where an old database dump is being restored -> the data must be from 2023 at the latest -> just a few seconds.

Reply:
> So, I tested the current approach on production data and it took only a few seconds.

On *your machine*, remember that in production the pods running these workloads are constrained in both memory and CPU (but mostly CPU for the web pod).

There are roughly 9k `FlawMeta` records of type `ACKNOWLEDGMENT` in production, so we'll have to perform a little under 9k `FlawAcknowledgment` INSERTs while going through Django's ORM, I honestly don't know whether this would be an issue or not but I can tell you that we've had some issues with this in the past.

That said I just checked and the migration deadline is 30min, which should be plenty I think, but in the future we should always have performance in mind when developing data migrations, it's essential that both time and memory constraints are kept in check

Reply:
Aha! Thank you for the context and explanation!

In that case, I'll _try_ to rework it as per your suggestion and learn to make it use reasonably-sized batches. Seems this is a skill that will be necessary in the future. If I fail, we can merge _this_ migration once it becomes necessary, but we won't have that comfort in the future.

Reply:
Fixed by using a bulk approach recommended by django. Added explanation of that approach into the docstring.

Tested locally with millions of instances and expected memory consumption is roughly under 100 MB RAM and done within a minute for 10k objects. Likely my estimate is overshot by an order of magnitude based on testing of the migration file with actual production data (a few seconds, memory consumption not measurable).